File: touyou/GitMe/GitMe/Domain/Main/MainConverter.swift
//
//  MainConverter.swift
//  GitMe
//
//  Created by 藤井陽介 on 2018/01/27.
//  Copyright © 2018 touyou. All rights reserved.
//

import UIKit
import RxSwift

// MARK: - MainConverterProtocol

protocol MainConverterProtocol {

    var isLoggedIn: Bool { get }

    func fetchEvent(at page: Int, every perPage: Int) -> Observable<[EventCellViewModel]>
    func fetchLoginUserInfo() -> Observable<UserInfoViewModel>
}

// MARK: - MainConverter

class MainConverter {

    // MARK: Private

    private let api = GitHubAPI.shared

    private func convertEventTitle(of event: Event) -> NSMutableAttributedString {

        let mutableAttributedString = NSMutableAttributedString()
        mutableAttributedString.append([(event.actor.displayLogin, .bold)])
        mutableAttributedString.append(createMutableTemplate(of: event))
        return mutableAttributedString
    }

    /// イベントに応じてメッセージをつくる。
    private func createMutableTemplate(of event: Event) -> [(String, NSMutableAttributedString.GitMeStyle)] {

        switch event.type {
        case .commitComment:

            return [(" \(event.payload.action!) comment \"\(event.payload.comment!.body)\"", .normal)]
        case .create:

            if let ref = event.payload.ref {

                return [(" created \(event.payload.refType!) ", .normal), (ref, .bold)]
            } else {

                return [(" created repository", .normal)]
            }
        case .delete:

            if let ref = event.payload.ref {

                return [(" deleted \(event.payload.refType!) ", .normal), (ref, .bold)]
            } else {

                return [(" deleted repository", .normal)]
            }
        case .fork:

            return [(" forked ", .normal), (event.payload.forkee!.fullName!, .bold)]
        case .gollum:

            return [(" did wiki page action", .normal)]
        case .installation:

            return [(" \(event.payload.action!) installation", .normal)]
        case .installationRepositories:

            return [(" \(event.payload.action!) installation", .normal)]
        case .issueComment:

            return [(" \(event.payload.action!) comment on issue ", .normal), ("#\(event.payload.issue!.number)", .bold)]
        case .issues:

           return [(" \(event.payload.action!) issue ", .normal), ("#\(event.payload.issue!.number)", .bold)]
        case .marketplacePurchase:

            return [(" \(event.payload.action!) marketplace", .normal)]
        case .member:

            return [(" \(event.payload.action!) ", .normal), (event.payload.member!.login, .bold)]
        case .orgBlock:

            return [(" \(event.payload.action!) ", .normal), (event.payload.member!.login, .bold)]
        case .projectCard:

            return [(" \(event.payload.action!) project card", .normal)]
        case .projectColumn:

            return [(" \(event.payload.action!) project column", .normal)]
        case .project:

            return [(" \(event.payload.action!) project", .normal)]
        case .public:

            return [(" made public", .normal)]
        case .pullRequest:

            return [(" \(event.payload.action!) pull request ", .normal), ("#\(event.payload.pullRequest!.number)", .bold)]
        case .pullRequestReview:

            return [(" \(event.payload.action!) review on pull request ", .normal), ("#\(event.payload.pullRequest!.number)", .bold)]
        case .pullRequestReviewComment:

            return [(" \(event.payload.action!) review comment on pull request ", .normal), ("#\(event.payload.pullRequest!.number)", .bold)]
        case .push:

            return [(" pushed to ", .normal), (event.payload.ref!.split(separator: "/").dropFirst(2).joined(separator: "/"), .bold)]
        case .release:

            return [(" \(event.payload.action!) \(event.payload.release!.tagName)", .normal)]
        case .watch:

            return [(" starred", .normal)]
        default:

            // ドキュメントを見る限りこれ意外はタイムラインとしては流れてこないということになっているはずなので
            return [(" created wrong event.", .normal)]
        }
    }

    private func convertRepositoryInfo(of repo: Repository) -> String {

        let starCount = repo.stargazersCount ?? 0
        let starString = starCount >= 1000 ? "\(starCount / 1000)K" : "\(starCount)"
        if let dateString = repo.updatedAt?.dateString {

            return "\(repo.language ?? "None")   ★\(starString)  Updated \(dateString)"
        }

        return "\(repo.language ?? "None")   ★\(starString)"
    }
}

// MARK: - Protocol Interface

extension MainConverter: MainConverterProtocol {

    var isLoggedIn: Bool {

        return api.isLoggedIn
    }

    func fetchEvent(at page: Int, every perPage: Int) -> Observable<[EventCellViewModel]> {

        return api.fetchEvents(at: page, every: perPage).map { events in

            events.map { event in

                let eventCellViewModel = EventCellViewModel()
                eventCellViewModel.iconUrl = event.actor.avatarUrl
                eventCellViewModel.eventTitle = self.convertEventTitle(of: event)
                eventCellViewModel.createAt = event.createdAt
                eventCellViewModel.repositoryName = event.repo?.name
                let (repo, readme) = self.api.fetchRepositoryInfo(of: event.repo!.url)
                eventCellViewModel.repoObservable = repo.map { (repository: Repository) in

                    let repoViewModel = RepositoryViewModel()
                    repoViewModel.repositoryDescription = repository.description
                    repoViewModel.repoInfo = self.convertRepositoryInfo(of: repository)
                    return repoViewModel
                }.share()
                eventCellViewModel.readmeObservable = readme.map { readme in

                    return readme.downloadUrl
                }.share()

                return eventCellViewModel
            }
        }
    }

    func fetchLoginUserInfo() -> Observable<UserInfoViewModel> {

        return api.logIn().map { user in

            let userInfoViewModel = UserInfoViewModel()
            userInfoViewModel.userName = user.login
            userInfoViewModel.iconUrl = user.avatarUrl
            return userInfoViewModel
        }
    }
}

File: touyou/GitMeAndroid/app/src/main/java/com/dev/touyou/gitme/Domain/Main/MainConverter.kt
package com.dev.touyou.gitme.Domain.Main

import com.dev.touyou.gitme.Data.Event
import com.dev.touyou.gitme.Data.EventType
import com.dev.touyou.gitme.Data.GitHubAPI
import com.dev.touyou.gitme.Data.Repository
import com.dev.touyou.gitme.Domain.UserInfoViewModel
import io.reactivex.Observable

/**
 * Created by touyou on 2018/02/26.
 */

interface MainConverterInterface {

    fun fetchEvent(page: Int, perPage: Int): Observable<Array<MainViewModel>>
    fun fetchLoginUserInfo(): Observable<UserInfoViewModel>
}

class MainConverter: MainConverterInterface {

    private fun convertEventTitle(event: Event): String {

        val eventTitle = event.actor.displayLogin + when(event.type) {
            EventType.Create -> " created ${event.repo!!.name}"
            EventType.IssueComment -> " ${event.payload.action!!} comment on issue #${event.payload.issue!!.number} in ${event.repo!!.name}"
            EventType.Issues -> " ${event.payload.action!!} issue #${event.payload.issue!!.number} in ${event.repo!!.name}"
            EventType.Push -> " pushed to ${event.payload.ref!!.split("/").last()} at ${event.repo!!.name}"
            EventType.Watch -> " starred ${event.repo!!.name}"
            else -> "${event.type.rawValue}: Unknown"
        }

        return eventTitle
    }

    private fun convertRepositoryInfo(repo: Repository): String {

        val starCount = repo.stargazersCount ?: 0
        val starString = if (starCount >= 1000) "${starCount / 1000}K" else starCount.toString()

        repo.updatedAt?.let {

            // TODO: 日付の変換形式を整える
            return "${repo.language ?: "None"}  ★$starString Updated $it"
        }

        return "${repo.language ?: "None"}  ★$starString"
    }

    /// MainConverterInterface

    override fun fetchEvent(page: Int, perPage: Int): Observable<Array<MainViewModel>> {

        return GitHubAPI.fetchEvents(page, perPage).map {

            it.map {

                val mainViewModel = MainViewModel()
                mainViewModel.iconUrl = it.actor.avatarUrl
                mainViewModel.eventTitle = convertEventTitle(it)
                mainViewModel.createAt = it.createdAt
                mainViewModel.repositoryName = it.repo?.name
                val (repo, readme) = GitHubAPI.fetchRepositoryInfo(it.repo!!.url)
                mainViewModel.repoObservable = repo.map {

                    val repoViewModel = RepositoryViewModel()
                    repoViewModel.repositoryDescription = it.description
                    repoViewModel.repoInfo = convertRepositoryInfo(it)
                    repoViewModel
                }
                mainViewModel.readmeObservable = readme.map {

                    it.downloadUrl
                }
                mainViewModel
            }.toTypedArray()
        }
    }

    override fun fetchLoginUserInfo(): Observable<UserInfoViewModel> {

        return GitHubAPI.logIn().map {

            val userInfoViewModel = UserInfoViewModel()
            userInfoViewModel.userName = it.login
            userInfoViewModel.iconUrl = it.avatarUrl
            userInfoViewModel
        }
    }
}